ethmactool: new package for getting HW MAC address or generating from CPU SN

kszaq opened the pull request:
Ethernet MAC address should be passed by u-boot in device tree. In many cases this address is not correct: `ethmactool` package allows userspace to take care of it.

Currently these are possible sources of "real" MAC:
 - Amlogic: cmdline, eFuse, CPU SN
 - default: CPU SN

For now the script is only triggered for `eth0` as we don't want to replace MAC on USB ethernet adapters or WiFi adapters.

MilhouseVH commented on packages/sysutils/ethmactool/scripts/ethmactool-config:
```diff
@@ -0,0 +1,73 @@
+#!/bin/bash
```
```suggestion
#!/bin/sh
```
as busybox `ash` is closer to `sh` than `bash`, I think?

MilhouseVH commented on packages/sysutils/ethmactool/scripts/ethmactool-config:
```diff
@@ -0,0 +1,73 @@
+MAC_STEP=""
+
+validate_mac() {
+  [ ${#MAC} = 12 -a "${MAC}" != "000000000000" ]
```
```suggestion
  [ ${#MAC} -eq 12 -a "${MAC}" != "000000000000" ]
```
as `${#MAC}` is a numeric result and shell should use `-eq` for numeric comparisons.

MilhouseVH commented:
What's the trigger to include this in a build - presumably there will be a dependency that includes this only for Rockchip and Amlogic* (project options?) as we wouldn't want to include this for RPi/Generic/etc as the udev rule will execute and log errors for those systems.

chewitt commented:
It needs to be added as an additional package in a project/device options file. If ends up being used more widely it should be promoted to a separate configurable item. For now the issue is mostly (not exclusively, but mostly) seen with Android box devices.

MilhouseVH commented:
Yep figured as much, but wasn't sure why the options haven't been included in the PR. Anyway as long as we only install this for devices that need it then this is all looks fine to me!

chewitt commented:
It's mostly required for Amlogic after the mainline branch gets merged into master, as that's where we see the bulk of the issues. This PR is a superset of ideas that we've been poking at for a while in test images. The refactoring is new though, so It needs to be road-tested a bit before merge.

Kwiboo commented:
You can remove the rockchip specific parts (or make it default), ethaddr and ethaddr1 is set  based on cpu sn or eeprom in vendor u-boot and we should only move to mainline u-boot if/when it does the same. Using android or stock u-boot is not supported on rockchip.

kszaq commented:
@MilhouseVH Thank you, all applied. 

@Kwiboo I made the `from_cpu_sn` step default, in that case we simply won't use the package for Rockchip.

CvH commented on packages/sysutils/ethmactool/package.mk:
```diff
@@ -0,0 +1,15 @@
+PKG_VERSION="1.0"
+PKG_ARCH="any"
+PKG_LICENSE="GPLv2"
+PKG_SECTION="system"
```
`PKG_SECTION` can be dropped

CvH commented on packages/sysutils/ethmactool/scripts/ethmactool-config:
```diff
@@ -0,0 +1,72 @@
+
+fixup_self_mac() {
+  # clear multicast bit and set local assignment bit (IEEE802)
+  MAC=$(printf '%X\n' "$(( (0x$MAC & 0xFEFFFFFFFFFF) | 0x020000000000 ))")
```
`$(( (0x$MAC & 0xFEFFFFFFFFFF) | 0x020000000000 ))"`
is that `"$(( ... ))"`needed ?

kszaq replied:
Yes, required for the calculation.

CvH commented on packages/sysutils/ethmactool/package.mk:
```diff
@@ -0,0 +1,15 @@
+
+PKG_NAME="ethmactool"
+PKG_VERSION="1.0"
+PKG_ARCH="any"
```
`PKG_ARCH="any"` can also be dropped and serves no purpose

vpeter4 commented on packages/sysutils/ethmactool/scripts/ethmactool-config:
```diff
@@ -0,0 +1,72 @@
+}
+
+from_cpu_sn() {
+  MAC=$(cat /proc/cpuinfo 2>/dev/null | awk '/Serial/ {print $3}' | cut -b-12)
```
Instead of final cut command awk's substring command could be used `awk '/Serial/ {print substr($3,1,12)}'`.

kszaq replied:
Thank you, will apply!

kszaq commented:
Force-pushed with the following changes:
* reverted to systemd service and make it start before `connman` to make sure `connman` picks the "new" MAC address for saving settings
* fixed `fixup_self_mac` to pad result with zeroes

Kwiboo commented on packages/sysutils/ethmactool/system.d/ethmactool-config.service:
```diff
@@ -0,0 +1,10 @@
+Before=connman.target
+
+[Service]
+Type=oneshot
```
I think a `RemainAfterExit=yes` should be insterted here

Kwiboo commented on packages/sysutils/ethmactool/system.d/ethmactool-config.service:
```diff
@@ -0,0 +1,10 @@
+[Unit]
+Description=Configure eth0 MAC address
+Before=connman.target
```
Possible also add a `After=sys-subsystem-net-devices-eth0.device` here (completely untested)

MilhouseVH commented:
The script and package looks fine, I'm not really a systemd expert so once @Kwiboo is happy I think this can merge.

kszaq commented:
Applied systemd service changes requested by @Kwiboo and modified to `Before=connman.service`
